README.md
# Pigpio Encoder

Version: 0.2.4
### Requires Python 3

#### Python module for the KY040 rotary encoder.
This module has been developed for quickly interface a rotary encoder with Raspberry Pi.
It's based on the [pigpio library](http://abyz.me.uk/rpi/pigpio/python.html) (cause it proved to be faster than rpi.GPIO or gpiozero libraries) so you need to install pigpio library and run pigpio daemon before starting your script.

## Features
- Easy to setup callback functions for the Rotary Encoder and the Switch.
- The Rotary Encoder has customizable min and max values (default 0-100).
- The Rotary Encoder increase/decrease value is customizable (default 1).
- The Switch can be activated or not.
- The Switch can have two different functions for short press or long press.
- Both Rotary and Switch have a customizable debounce value (default 300ms)

## Installation
- Install the pigpio library *(check [pigpio documentation](http://abyz.me.uk/rpi/pigpio/download.html) for alternative installation method)*
 - `sudo apt-get update`
 - `sudo apt-get install pigpio python-pigpio python3-pigpio`
- Install the pigpio_encoder library
 - `python3 -mpip install pigpio_encoder` (consider add --user option)
- start pigpio daemon
 - `sudo pigpiod`

## How to use
- import the module
    ```python
    from pigpio_encoder.rotary import Rotary
    ```
- create a callback function for the Rotary Encoder counter.
    > You must pass a positional argument to retrieve the counter value.

    ```python
    def rotary_callback(counter):
        # some action with counter...
    ```

- create a callback function for Up-Rotation events.

    ```python
    def up_callback():
        # some action if rotated upward
    ```

- create a callback function for Down-Rotation events.

    ```python
    def down_callback():
        # some action if rotated downward
    ```

- create callbacks functions for the Switch
    > If you intend to use the switch you must create at least the "short press" callback. The "long press" callback is necessary if you want to use that feature.

    ```python
    def sw_short_callback():
        # some action...
    ```
    ```python
    def sw_long_callback():
        # some action...
      ```
- create the rotary object
    > here you setup the gpio id as keyword argument. If you don't pass the switch parameter the switch won't be activated.

    ```python
    my_rotary = Rotary(
            clk_gpio=<gpio_id of clk signal>,
            dt_gpio=<gpio_id of dt signal>,
            sw_gpio=<gpio_id of switch signal>
            )
    ```
- setup the rotary encoder for counting
    > here you can setup min and max values for the encoder, the increase/decrease value, a debounce value (default 300ms) and the callback function.

    ```python
    my_rotary.setup_rotary(
            min=<min_value>,
            max=<max_value>,
            scale=<scale_value>,
            debounce=<debounce_value>,
            rotary_callback=<rotary_callback>
            )
    ```

- Optional setup for up and down rotation events
    > here you can setup min and max values for the encoder, the increase/decrease value, a debounce value (default 300ms) and the callback functions.

    ```python
    my_rotary.setup_rotary(
            min=<min_value>,
            max=<max_value>,
            scale=<scale_value>,
            debounce=<debounce_value>,
            up_callback=<up_callback>
            down_callback=<down_callback>
            )
    ```


- setup the switch
    > if you have specified the switch pin when creating the encoder object, here you can setup the debounce value, the long press option and the callbacks.

    ```python
    my_rotary.setup_switch(
            debounce=<debounce_value>,
            long_press=<True>,
            sw_short_callback=<sw_short_callback>,
            sw_long_callback=<sw_long_callback>
            )
    ```

- start the listener
    ```python
    my_rotary.watch()
    ```
  Please note that calling .watch() this is only for demostration purposes. It is not neccessary for the encoder to operate it can even cause problems.


___
#### Basic example using default values

```python
from pigpio_encoder.rotary import Rotary


def rotary_callback(counter):
  print("Counter value: ", counter)


def sw_short():
  print("Switch pressed")


def up_callback():
  print("Up rotation")


def down_callback():
  print("Down rotation")


my_rotary = Rotary(clk_gpio=27, dt_gpio=22, sw_gpio=17)
my_rotary.setup_rotary(
  rotary_callback=rotary_callback,
  up_callback=up_callback,
  down_callback=down_callback,
)
my_rotary.setup_switch(sw_short_callback=sw_short)

my_rotary.watch()
```
  Please note that calling .watch() this is only for demostration purposes. It is not neccessary for the encoder to operate it can even cause problems.


___

#### Example using all the Features

```python
from pigpio_encoder.rotary import Rotary


def rotary_callback(counter):
  print("Counter value: ", counter)


def sw_short():
  print("Switch short press")


def sw_long():
  print("Switch long press")


my_rotary = Rotary(
  clk_gpio=27,
  dt_gpio=22,
  sw_gpio=17
)
my_rotary.setup_rotary(
  min=10,
  max=300,
  scale=5,
  debounce=200,
  rotary_callback=rotary_callback
)
my_rotary.setup_switch(
  debounce=200,
  long_press=True,
  sw_short_callback=sw_short,
  sw_long_callback=sw_long
)

my_rotary.watch()

```
  Please note that calling .watch() this is only for demostration purposes. It is not neccessary for the encoder to operate it can even cause problems.

___

## Thanks to...
- [joan2937](https://github.com/joan2937) for the awesome [pigpio library](https://github.com/joan2937/pigpio)
- [Raphael Yancey](https://github.com/raphaelyancey) for inspiring me this library with his [similar project](https://github.com/raphaelyancey/pyKY040)


CHANGELOG.md
# Change Log

## 0.2.3

### Added

* Wrap around counter feature. [#15](https://github.com/vash3d/pigpio_encoder/issues/15) 

## 0.2.4

### Added

* Version for Micropython
* Fixed broken deployment [#16](https://github.com/vash3d/pigpio_encoder/issues/16)
* Added a changelog file CHANGELOG.md
* Added poetry support


src/test.py
import unittest
from unittest.mock import patch, Mock

import sys
sys.modules['pigpio'] = 'huhu'

from pigpio_encoder import Rotary

class TestSetupSimple(unittest.TestCase):

    @patch('pigpio_encoder.pigpio')
    def test(self, pigpio):

        def rotary_callback(counter):
            print("Counter value: ", counter)

        def sw_short():
            print("Switch pressed")

        my_rotary = Rotary(clk_gpio=27, dt_gpio=22, sw_gpio=17)
        my_rotary.setup_rotary(rotary_callback=rotary_callback)
        my_rotary.setup_switch(sw_short_callback=sw_short)


class TestSetupAll(unittest.TestCase):

    @patch('pigpio_encoder.pigpio')
    def test(self, pigpio):
        def rotary_callback(counter):
            print("Counter value: ", counter)

        def sw_short(self):
            print("Switch short press")

        def sw_long(self):
            print("Switch long press")

        my_rotary = Rotary(clk_gpio=27, dt_gpio=22, sw_gpio=17)
        my_rotary.setup_rotary(min=10, max=300, scale=5, debounce=200, rotary_callback=rotary_callback)
        my_rotary.setup_switch(debounce=200, long_press=True, sw_short_callback=sw_short, sw_long_callback=sw_long)


if __name__ == '__main__':
    unittest.main()


src/pigpio_encoder/rotary.py
# Rotary encoder class based on pigpio library
# version: 0.2.5

try:
    import pigpio
except ModuleNotFoundError:
    import sys
    from unittest.mock import MagicMock
    sys.modules['pigpio'] = MagicMock()


import time

# States
dt_gpio1 = 'D'  # dt_gpio is high
dt_gpio0 = 'd'  # dt_gpio is low
clk_gpio1 = 'C'  # clk_gpio is high
clk_gpio0 = 'c'  # clk_gpio is low

# State sequences
SEQUENCE_UP = dt_gpio1 + clk_gpio1 + dt_gpio0 + clk_gpio0
SEQUENCE_DOWN = clk_gpio1 + dt_gpio1 + clk_gpio0 + dt_gpio0


class Rotary:

    sequence = ''

    # Default values for the rotary encoder
    min = 0
    max = 100
    scale = 1
    debounce = 300
    _counter = 0
    last_counter = 0
    rotary_callback = None

    # Default values for the sw_gpioitch
    sw_gpio_debounce = 300
    long_press_opt = False
    sw_gpio_short_callback = None
    sw_gpio_long_callback = None
    up_callback = None
    down_callback = None
    sw_short_callback = None
    sw_long_callback = None
    sw_debounce = None

    wait_time = time.time()
    long = False

    def __init__(self, clk_gpio=None, dt_gpio=None, sw_gpio=None):
        if not (clk_gpio and dt_gpio):
            raise BaseException("clk_gpio and dt_gpio pin must be specified!")

        self.pi = pigpio.pi()
        self.clk_gpio = clk_gpio
        self.dt_gpio = dt_gpio
        self.pi.set_glitch_filter(self.clk_gpio, self.debounce)
        self.pi.set_glitch_filter(self.dt_gpio, self.debounce)
        if sw_gpio is not None:
            self.sw_gpio = sw_gpio
            self.pi.set_pull_up_down(self.sw_gpio, pigpio.PUD_UP)
            self.pi.set_glitch_filter(self.sw_gpio, self.sw_gpio_debounce)
        self.setup_pigpio_callbacks()

    def setup_pigpio_callbacks(self):
        self.pi.callback(self.clk_gpio, pigpio.FALLING_EDGE, self.clk_gpio_fall)
        self.pi.callback(self.clk_gpio, pigpio.RISING_EDGE, self.clk_gpio_rise)
        self.pi.callback(self.dt_gpio, pigpio.FALLING_EDGE, self.dt_gpio_fall)
        self.pi.callback(self.dt_gpio, pigpio.RISING_EDGE, self.dt_gpio_rise)
        if self.sw_gpio is not None:
            self.pi.callback(self.sw_gpio, pigpio.FALLING_EDGE, self.sw_gpio_fall)
            self.pi.callback(self.sw_gpio, pigpio.RISING_EDGE, self.sw_gpio_rise)

    @property
    def counter(self):
        return self._counter

    @counter.setter
    def counter(self, value):
        if self._counter != value:
            self._counter = value
            if self.rotary_callback:
                self.rotary_callback(self._counter)

    def clk_gpio_fall(self, _gpio, _level, _tick):
        if len(self.sequence) > 2:
            self.sequence = ''
        self.sequence += clk_gpio1

    def clk_gpio_rise(self, _gpio, _level, _tick):
        self.sequence += clk_gpio0
        if self.sequence == SEQUENCE_UP:
            if self.counter < self.max:
                self.counter += self.scale
            if self.up_callback:
                self.up_callback(self._counter)
            self.sequence = ''

    def dt_gpio_fall(self, _gpio, _level, _tick):
        if len(self.sequence) > 2:
            self.sequence = ''
        self.sequence += dt_gpio1

    def dt_gpio_rise(self, _gpio, _level, _tick):
        self.sequence += dt_gpio0
        if self.sequence == SEQUENCE_DOWN:
            if self.counter > self.min:
                self.counter -= self.scale
            if self.down_callback:
                self.down_callback(self._counter)
            self.sequence = ''

    def sw_gpio_rise(self, _gpio, _level, _tick):
        if self.long_press_opt:
            if not self.long:
                self.short_press()

    def sw_gpio_fall(self, _gpio, _level, _tick):
        if self.long_press_opt:
            self.long = False
            press_time = time.time()
            while self.pi.read(self.sw_gpio) == 0:
                self.wait_time = time.time()
                time.sleep(0.1)
                if self.wait_time - press_time > 1.5:
                    self.long_press()
                    self.long = True
                    break
        else:
            self.short_press()

    def setup_rotary(
            self,
            rotary_callback=None,
            up_callback=None,
            down_callback=None,
            min=None,
            max=None,
            scale=None,
            debounce=None,
         ):
        if not (rotary_callback or up_callback or down_callback):
            print('At least one callback should be given')
        # rotary callback has to be set first since the self.counter property depends on it
        self.rotary_callback = rotary_callback
        self.up_callback = up_callback
        self.down_callback = down_callback
        if min is not None:
            self.min = min
            self.counter = self.min
            self.last_counter = self.min
        if max is not None:
            self.max = max
        if scale is not None:
            self.scale = scale
        if debounce is not None:
            self.debounce = debounce
            self.pi.set_glitch_filter(self.clk_gpio, self.debounce)
            self.pi.set_glitch_filter(self.dt_gpio, self.debounce)

    def setup_switch(self,
                     sw_short_callback=None,
                     sw_long_callback=None,
                     debounce=None,
                     long_press=None
                     ):
        assert sw_short_callback is not None or sw_long_callback is not None
        if sw_short_callback is not None:
            self.sw_short_callback = sw_short_callback
        if sw_long_callback is not None:
            self.sw_long_callback = sw_long_callback
        if debounce is not None:
            self.sw_debounce = debounce
            self.pi.set_glitch_filter(self.sw_gpio, self.sw_debounce)
        if long_press is not None:
            self.long_press_opt = long_press

    @staticmethod
    def watch():
        """
        A simple convenience function to have a waiting loop
        """
        while True:
            time.sleep(10)

    def short_press(self):
        self.sw_short_callback()

    def long_press(self):
        self.sw_long_callback()


src/pigpio_encoder/debounce.py
from machine import Timer


def debounce(wait):
    """Postpone a functions execution until after some time has elapsed

    :type wait: int
    :param wait: The amount of Seconds to wait before the next call can execute.
    """

    def decorator(fun):
        class Debounced():
            tid = 10

            def __init__(self):
                self.run = True
                self.tid = Debounced.tid
                Debounced.tid += 1

            def __call__(self, *args, **kwargs):
                self.args = args
                self.kwargs = kwargs

                if self.run:
                    self.call_it()

                self.run = False
                self.t = Timer(5)
                self.t.init(mode=Timer.ONE_SHOT, period=wait, callback=self.clear_run)

            def call_it(self):
                fun(*self.args, **self.kwargs)

            def clear_run(self, dummy):
                self.run = True

        return Debounced()

    return decorator


src/pigpio_encoder/rotary_mp.py
# Rotary encoder class for ESP32 with micropython
# version: 0.2.5

from machine import Pin, Timer
import micropython

import time

# States
dt_gpio1 = 'D'  # dt_gpio is high
dt_gpio0 = 'd'  # dt_gpio is low
clk_gpio1 = 'C'  # clk_gpio is high
clk_gpio0 = 'c'  # clk_gpio is low

# State sequences
SEQUENCE_UP = dt_gpio1 + clk_gpio1 + dt_gpio0 + clk_gpio0
SEQUENCE_DOWN = clk_gpio1 + dt_gpio1 + clk_gpio0 + dt_gpio0

timer_init = lambda t, p, cb: t.init(period=p, callback=cb)




class Rotary:

    sequence = ''

    # Default values for the rotary encoder
    min = 0
    max = 100
    scale = 1
    debounce = 300
    _counter = 0
    last_counter = 0
    rotary_callback = None

    # Default values for the sw_gpioitch
    sw_gpio_debounce = 300
    long_press_opt = False
    sw_gpio_short_callback = None
    sw_gpio_long_callback = None
    up_callback = None
    down_callback = None
    sw_short_callback = None
    sw_long_callback = None
    sw_debounce = None

    wait_time = time.time()
    long = False

    def __init__(self, clk_gpio=None, dt_gpio=None, sw_gpio=None):
        if not (clk_gpio and dt_gpio):
            raise BaseException("clk_gpio and dt_gpio pin must be specified!")

        self.clk_gpio = clk_gpio
        self.dt_gpio = dt_gpio
        if sw_gpio is not None:
            self.sw_gpio = sw_gpio
        self.setup_pigpio_callbacks()

    def setup_pigpio_callbacks(self):
        clk_Pin = Pin(self.clk_gpio, Pin.IN, Pin.PULL_UP)
        self.clk_state = clk_Pin.value()
        dt_Pin = Pin(self.dt_gpio, Pin.IN, Pin.PULL_UP)
        self.dt_state = dt_Pin.value()
        sw_Pin = Pin(self.sw_gpio, Pin.IN, Pin.PULL_UP)
        self.sw_state = sw_Pin.value()
        
        clk_Pin.irq(handler=self.clk_gpio_call, trigger=Pin.IRQ_FALLING | Pin.IRQ_RISING)
        dt_Pin.irq(handler=self.dt_gpio_call, trigger=Pin.IRQ_FALLING | Pin.IRQ_RISING)
        
        if self.sw_gpio is not None:
            sw_Pin.irq(handler=self.sw_gpio_call, trigger=Pin.IRQ_FALLING | Pin.IRQ_RISING)

    @property
    def counter(self):
        return self._counter

    @counter.setter
    def counter(self, value):
        if self._counter != value:
            self._counter = value
            if self.rotary_callback:
                self.rotary_callback(self._counter)

    def clk_gpio_call(self, pin):
        val = pin.value()
        if val < self.clk_state:
            self.clk_gpio_fall()
        elif val > self.clk_state:
            self.clk_gpio_rise()
        self.clk_state = val

    def clk_gpio_fall(self ):
#        print('clk_gpio_fall')
        if len(self.sequence) > 2:
            self.sequence = ''
        self.sequence += clk_gpio1

    def clk_gpio_rise(self): 
#        print('clk_gpio_rise')
        self.sequence += clk_gpio0
        if self.sequence == SEQUENCE_UP:
            if self.counter < self.max:
                self.counter += self.scale
            if self.up_callback:
                self.up_callback(self._counter)
            self.sequence = ''

    def dt_gpio_call(self, pin):
        val = pin.value()
        if val < self.dt_state:
            self.dt_gpio_fall()
        elif val > self.dt_state:
            self.dt_gpio_rise()
        self.dt_state = val

    def dt_gpio_fall(self):
#        print('dt_gpio_fall')
        if len(self.sequence) > 2:
            self.sequence = ''
        self.sequence += dt_gpio1

    def dt_gpio_rise(self):
#        print('dt_gpio_rise')
        self.sequence += dt_gpio0
        if self.sequence == SEQUENCE_DOWN:
            if self.counter > self.min:
                self.counter -= self.scale
            if self.down_callback:
                self.down_callback(self._counter)
            self.sequence = ''

    def sw_gpio_call(self, pin):
        val = pin.value()
        if val < self.sw_state:
            self.sw_gpio_fall()
        elif val > self.sw_state:
            self.sw_gpio_rise()
        self.sw_state = val

    def sw_gpio_rise(self):
#        print('sw_gpio_rise')
        if self.long_press_opt:
            if not self.long:
                self.short_press()

    def sw_gpio_fall(self):
#        print('sw_gpio_fall')
        if self.long_press_opt:
            self.long = False
            press_time = time.time()
            while self.pi.read(self.sw_gpio) == 0:
                self.wait_time = time.time()
                time.sleep(0.1)
                if self.wait_time - press_time > 1.5:
                    self.long_press()
                    self.long = True
                    break
        else:
            self.short_press()

    def setup_rotary(
            self,
            rotary_callback=None,
            up_callback=None,
            down_callback=None,
            min=None,
            max=None,
            scale=None,
            debounce=None,
         ):
        if not (rotary_callback or up_callback or down_callback):
            print('At least one callback should be given')
        # rotary callback has to be set first since the self.counter property depends on it
        self.rotary_callback = rotary_callback
        self.up_callback = up_callback
        self.down_callback = down_callback
        if min is not None:
            self.min = min
            self.counter = self.min
            self.last_counter = self.min
        if max is not None:
            self.max = max
        if scale is not None:
            self.scale = scale

    def setup_switch(self,
                     sw_short_callback=None,
                     sw_long_callback=None,
                     debounce=None,
                     long_press=None
                     ):
        assert sw_short_callback is not None or sw_long_callback is not None
        if sw_short_callback is not None:
            self.sw_short_callback = sw_short_callback
        if sw_long_callback is not None:
            self.sw_long_callback = sw_long_callback
        if debounce is not None:
            self.sw_debounce = debounce
            self.pi.set_glitch_filter(self.sw_gpio, self.sw_debounce)
        if long_press is not None:
            self.long_press_opt = long_press

    @staticmethod
    def watch():
        """
        A simple convenience function to have a waiting loop
        """
        while True:
            time.sleep(10)

    def short_press(self):
        self.sw_short_callback()

    def long_press(self):
        self.sw_long_callback()


src/pigpio_encoder.py
# Rotary encoder class based on pigpio library
# version: 0.2.5

try:
    import pigpio
except ModuleNotFoundError:
    import sys
    from unittest.mock import MagicMock
    sys.modules['pigpio'] = MagicMock()


import time

# States
dt_gpio1 = 'D'  # dt_gpio is high
dt_gpio0 = 'd'  # dt_gpio is low
clk_gpio1 = 'C'  # clk_gpio is high
clk_gpio0 = 'c'  # clk_gpio is low

# State sequences
SEQUENCE_UP = dt_gpio1 + clk_gpio1 + dt_gpio0 + clk_gpio0
SEQUENCE_DOWN = clk_gpio1 + dt_gpio1 + clk_gpio0 + dt_gpio0


class Rotary:

    sequence = ''

    # Default values for the rotary encoder
    min = 0
    max = 100
    scale = 1
    debounce = 300
    _counter = 0
    last_counter = 0
    rotary_callback = None

    # Default values for the sw_gpioitch
    sw_gpio_debounce = 300
    long_press_opt = False
    sw_gpio_short_callback = None
    sw_gpio_long_callback = None
    up_callback = None
    down_callback = None
    sw_short_callback = None
    sw_long_callback = None
    sw_debounce = None

    wait_time = time.time()
    long = False

    def __init__(self, clk_gpio=None, dt_gpio=None, sw_gpio=None):
        if not (clk_gpio and dt_gpio):
            raise BaseException("clk_gpio and dt_gpio pin must be specified!")

        self.pi = pigpio.pi()
        self.clk_gpio = clk_gpio
        self.dt_gpio = dt_gpio
        self.pi.set_glitch_filter(self.clk_gpio, self.debounce)
        self.pi.set_glitch_filter(self.dt_gpio, self.debounce)
        if sw_gpio is not None:
            self.sw_gpio = sw_gpio
            self.pi.set_pull_up_down(self.sw_gpio, pigpio.PUD_UP)
            self.pi.set_glitch_filter(self.sw_gpio, self.sw_gpio_debounce)
        self.setup_pigpio_callbacks()

    def setup_pigpio_callbacks(self):
        self.pi.callback(self.clk_gpio, pigpio.FALLING_EDGE, self.clk_gpio_fall)
        self.pi.callback(self.clk_gpio, pigpio.RISING_EDGE, self.clk_gpio_rise)
        self.pi.callback(self.dt_gpio, pigpio.FALLING_EDGE, self.dt_gpio_fall)
        self.pi.callback(self.dt_gpio, pigpio.RISING_EDGE, self.dt_gpio_rise)
        if self.sw_gpio is not None:
            self.pi.callback(self.sw_gpio, pigpio.FALLING_EDGE, self.sw_gpio_fall)
            self.pi.callback(self.sw_gpio, pigpio.RISING_EDGE, self.sw_gpio_rise)

    @property
    def counter(self):
        return self._counter

    @counter.setter
    def counter(self, value):
        if self._counter != value:
            self._counter = value
            if self.rotary_callback:
                self.rotary_callback(self._counter)

    def clk_gpio_fall(self, _gpio, _level, _tick):
        if len(self.sequence) > 2:
            self.sequence = ''
        self.sequence += clk_gpio1

    def clk_gpio_rise(self, _gpio, _level, _tick):
        self.sequence += clk_gpio0
        if self.sequence == SEQUENCE_UP:
            if self.up_callback:
                self.up_callback()
            if self.counter + self.scale <= self.max:
                self.counter += self.scale
            elif self.wrap_around:
                self.counter = (self.counter + self.scale) % (self.max + 1) + self.min
            self.sequence = ''

    def dt_gpio_fall(self, _gpio, _level, _tick):
        if len(self.sequence) > 2:
            self.sequence = ''
        self.sequence += dt_gpio1

    def dt_gpio_rise(self, _gpio, _level, _tick):
        self.sequence += dt_gpio0
        if self.sequence == SEQUENCE_DOWN:
            if self.down_callback:
                self.down_callback()
            if self.counter - self.scale >= self.min:
                self.counter -= self.scale
            elif self.wrap_around:
                self.counter = (self.counter - self.min - self.scale) % (self.max + 1)
            self.sequence = ''

    def sw_gpio_rise(self, _gpio, _level, _tick):
        if self.long_press_opt:
            if not self.long:
                self.short_press()

    def sw_gpio_fall(self, _gpio, _level, _tick):
        if self.long_press_opt:
            self.long = False
            press_time = time.time()
            while self.pi.read(self.sw_gpio) == 0:
                self.wait_time = time.time()
                time.sleep(0.1)
                if self.wait_time - press_time > 1.5:
                    self.long_press()
                    self.long = True
                    break
        else:
            self.short_press()

    def setup_rotary(
            self,
            rotary_callback=None,
            up_callback=None,
            down_callback=None,
            min=None,
            max=None,
            scale=None,
            debounce=None,
            wrap_around=None,
         ):
        if not (rotary_callback or up_callback or down_callback):
            print('At least one callback should be given')
        # rotary callback has to be set first since the self.counter property depends on it
        self.rotary_callback = rotary_callback
        self.up_callback = up_callback
        self.down_callback = down_callback
        self.wrap_around = wrap_around
        if min is not None:
            self.min = min
            self.counter = self.min
            self.last_counter = self.min
        if max is not None:
            self.max = max
        if scale is not None:
            self.scale = scale
        if debounce is not None:
            self.debounce = debounce
            self.pi.set_glitch_filter(self.clk_gpio, self.debounce)
            self.pi.set_glitch_filter(self.dt_gpio, self.debounce)

    def setup_switch(self,
                     sw_short_callback=None,
                     sw_long_callback=None,
                     debounce=None,
                     long_press=None
                     ):
        assert sw_short_callback is not None or sw_long_callback is not None
        if sw_short_callback is not None:
            self.sw_short_callback = sw_short_callback
        if sw_long_callback is not None:
            self.sw_long_callback = sw_long_callback
        if debounce is not None:
            self.sw_debounce = debounce
            self.pi.set_glitch_filter(self.sw_gpio, self.sw_debounce)
        if long_press is not None:
            self.long_press_opt = long_press

    @staticmethod
    def watch():
        """
        A simple convenience function to have a waiting loop
        """
        while True:
            time.sleep(10)

    def short_press(self):
        self.sw_short_callback()

    def long_press(self):
        self.sw_long_callback()
